Benchmarks for elections-phragmen pallet

kianenigma opened the pull request:
Weights analysis should follow as soon as this is executed on the bench machine. 

What needs to be resolved first is the iteration counts. I am not sure what is reasonable to finish in a timely manner. I added notes to each benchmark to show my remarks and which complexity variables they each have. 

related https://github.com/paritytech/substrate/issues/5596

- [ ] make an issue in the apps side to ensure all the api changes are synced. 

I changed the label to client noteworthy because the api to work with elections will change here. This PR is basically more than just weight annotation, it also makes the pallet safe.

shawntabrizi commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -0,0 +1,283 @@
+	-> Result<(), &'static str>
+{
+	<Elections<T>>::vote(RawOrigin::Signed(caller).into(), votes, stake)
+		.map_err(|e| "failed to submit vote")
```
```suggestion
		.map_err(|_| "failed to submit vote")
```

shawntabrizi commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -0,0 +1,283 @@
+	}: _(RawOrigin::Signed(caller))
+
+	report_defunct_voter {
+		// has no complexity parameter. two voters exist. One only votes fro outdated candidates.
```
```suggestion
		// has no complexity parameter. two voters exist. One only votes for outdated candidates.
```

shawntabrizi commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -0,0 +1,283 @@
+	}: _(RawOrigin::Signed(caller), votes, stake)
+
+	remove_voter {
+		// No complexity parameter. while we can vote for numerous candidates and then remove them
+		// (and I think each can have slightly different performance), we cannot express this via
+		// the parameters. Hence, we only check for the median, 8 votes to be removed. Only use
+		// account seed as complexity parameter.
+		// TODO: what should be the upper bound here? the first benchmark runs 16 iterations.. this
+		// one 1000? These are used only to iterate and run the benchmark s times. They are not a
+		// parameter in any way.
+		let s in 0 .. 1000;
```
Technically this should be tested over the number of locks and the number of other voters in the system.

So you should populate the user you remove with something like 0 .. 50 locks, and populating the voters map (probably doing this in addition to us adding voters into the state initially with genesis)

shawntabrizi commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -0,0 +1,283 @@
+	_ {}
+
+	// -- Signed ones
+	vote {
```
A note on the code of the vote extrinsic, you could re-order some of the ensure checks so that the extrinsic fails sooner without reading into storage. for example these two checks:

```
			ensure!(votes.len() <= MAXIMUM_VOTE, Error::<T>::MaximumVotesExceeded);
			ensure!(!votes.is_empty(), Error::<T>::NoVotes);
```

should be the first two lines in your extrinsic, as it does not require any additional runtime operations to confirm.

kianenigma replied:
great recommendation!

shawntabrizi commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -0,0 +1,283 @@
+	// -- Signed ones
+	vote {
+		// range of candidates to vote for. Direct argument of the dispatchable.
+		let x in 1 .. (MAXIMUM_VOTE as u32);
```
`vote` extrinsic depends on the length of `candidates` and `members` right?

```
let candidates_count = <Candidates<T>>::decode_len().unwrap_or(0) as usize;
let members_count = <Members<T>>::decode_len().unwrap_or(0) as usize;
```

Also depends on the number of locks on a user, because you set a new lock.

kianenigma replied:
depends in which sense? you can vote a number `x` where `(candidates + members + runners)< x < MAXIMUM_VOTE `. And assuming the worse case, the bound is always `MAXIMUM_VOTE`, i.e. we have more candidates and members than `MAXIMUM_VOTE`. The user can then vote for any number `1 < x < MAXIMUM_VOTE`.

and I am not aware of how locking is related here. Indeed we set a new lock here. So it matters if my user had previously 1 locks or 100 locks?

kianenigma replied:
Okay I got the deal with locks. 

but members and candidates, I disagree. It does not matter if we have members, candidates, or runners-up. We can vote for at most `MAXIMUM_VOTE`, regardless of how many members etc. we have. The onle case is if sum of these 3 are less than `MAXIMUM_VOTE `, which I safely ignore.

shawntabrizi commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -0,0 +1,283 @@
+	}: _(RawOrigin::Signed(caller))
+
+	report_defunct_voter {
+		// has no complexity parameter. two voters exist. One only votes fro outdated candidates.
```
This is not accurate. You write in your code

```
			// Checking if someone is a candidate and a member here is O(LogN), making the whole
			// function O(MLonN) with N candidates in total and M of them being voted by `target`.
```

and `is_defunct_voter` iterates over `votes_of(who)`, so you should populate this vec using component paramters.

shawntabrizi replied:
is defunct voter also does a binary search over the candidates and members and runner. So you need to populate all of that

shawntabrizi commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -0,0 +1,283 @@
+		submit_voter::<T>(caller.clone(), all_candidates, stake)?;
+	}: _(RawOrigin::Signed(caller))
+
+	report_defunct_voter {
```
You need to have separate benchmarks to test for the `if` statement in this extrinsic.

where valid is true or false

kianenigma replied:
good point, added.

shawntabrizi commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -0,0 +1,283 @@
+	}: _(RawOrigin::Signed(account_2), account_1_lookup)
+
+	submit_candidacy {
+		// No complexity parameter. Use account seed to iterate.
```
`is_candidate`, `is_member`, is_runner` is all used here again, so you need to populate these vectors

shawntabrizi commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -0,0 +1,283 @@
+	}: _(RawOrigin::Signed(candidate_account.clone()))
+
+	renounce_candidacy {
+		// No complexity parameter. Use account seed to iterate.
```
This is certainly not true. `remove_and_replace_member` does a lot of things here that you need to populate.

shawntabrizi commented:
I have stopped reviewing this PR. In general, it seems like there is a misunderstanding of the params and complexity we need to test here.

In any and all situations that we iterate over some vector (which happens a lot here), we need to vary the number of elements in that vector, or set it to the max number of elements we expect in that vector. Otherwise, operations like a binary search or an iter take no time at all.

If we find that these components have significant impact on the complexity of the function, we need to force the user to tell us about those parameters so that we can accurately weight the function.

kianenigma commented:
Some initial results on my machine. I won't care about the absolute values. Just to see if parameters are good. 

<details>
<summary>vote</summary>
<pre>
Pallet: "elections", Extrinsic: "vote", Lowest values: [], Highest values: [], Steps: [], Repeat: 10
Median Slopes Analysis
========

Model:
Time ~=       34
    + x        0
    + l    0.166
              µs

Min Squares Analysis
========

Data points distribution:
    x     l   mean µs  sigma µs       %
    1    20     41.33      1.49    3.6%
    2    20     39.83     1.213    3.0%
    3    20     36.83     0.372    1.0%
    4    20     35.83     0.372    1.0%
    5    20        36         0    0.0%
    6    20      35.5       0.5    1.4%
    7    20     35.16     0.372    1.0%
    8    20        35         0    0.0%
    9    20     35.16     0.372    1.0%
   10    20      34.5       0.5    1.4%
   11    20     34.16     0.372    1.0%
   12    20      34.5       0.5    1.4%
   13    20     34.33     0.471    1.3%
   14    20      34.5       0.5    1.4%
   15    20        34         0    0.0%
   16     1        31         0    0.0%
   16     2      31.5       0.5    1.5%
   16     3     31.66     0.471    1.4%
   16     4     31.66     0.471    1.4%
   16     5        32         0    0.0%
   16     6        32         0    0.0%
   16     7     32.16     0.372    1.1%
   16     8     32.83     0.372    1.1%
   16     9        33         0    0.0%
   16    10     33.16     0.372    1.1%
   16    11        33         0    0.0%
   16    12     33.33     0.471    1.4%
   16    13     33.16     0.372    1.1%
   16    14     33.16     0.372    1.1%
   16    15     33.16     0.372    1.1%
   16    16     34.16     0.372    1.0%
   16    17        34         0    0.0%
   16    18     34.66     0.471    1.3%
   16    19        34         0    0.0%
   16    20        34         0    0.0%

Quality and confidence:
param     error
x         0.017
l         0.013

Model:
Time ~=    36.01
    + x        0
    + l    0.113
              µs
</pre>
</details>

<details>
<summary>remove_voter</summary>
<pre>
Median Slopes Analysis
========

Model:
Time ~=    25.98
    + c        0
    + l    0.055
              µs

Min Squares Analysis
========

Data points distribution:
    c     l   mean µs  sigma µs       %
    1    20      29.5       0.5    1.6%
    2    20        32     3.464   10.8%
    3    20        29     0.577    1.9%
    4    20      27.5     1.118    4.0%
    5    20      25.5       0.5    1.9%
    6    20     24.33     0.471    1.9%
    7    20      25.5       0.5    1.9%
    8    20        25         0    0.0%
    9    20        25         0    0.0%
   10    20     25.16     0.372    1.4%
   11    20     25.16     0.372    1.4%
   12    20        27         0    0.0%
   13    20     25.33     0.471    1.8%
   14    20     26.16     0.372    1.4%
   15    20     25.66     0.471    1.8%
   16     1     24.83     0.372    1.4%
   16     2     24.16     0.372    1.5%
   16     3     24.66     0.471    1.9%
   16     4      24.5       0.5    2.0%
   16     5     24.33     0.471    1.9%
   16     6     24.33     0.471    1.9%
   16     7        25         0    0.0%
   16     8        25         0    0.0%
   16     9      24.5       0.5    2.0%
   16    10     25.33     0.471    1.8%
   16    11        25         0    0.0%
   16    12        25         0    0.0%
   16    13        25         0    0.0%
   16    14      25.5       0.5    1.9%
   16    15        26         0    0.0%
   16    16     25.16     0.372    1.4%
   16    17        26         0    0.0%
   16    18        26         0    0.0%
   16    19     25.33     0.471    1.8%
   16    20      25.7     0.458    1.7%

Quality and confidence:
param     error
c         0.023
l         0.017

Model:
Time ~=    27.63
    + c        0
    + l    0.033
              µs
</pre>
</details>

shawntabrizi commented on bin/node/runtime/src/lib.rs:
```diff
@@ -928,6 +928,7 @@ impl_runtime_apis! {
 			add_benchmark!(params, batches, b"treasury", Treasury);
 			add_benchmark!(params, batches, b"utility", Utility);
 			add_benchmark!(params, batches, b"vesting", Vesting);
+			add_benchmark!(params, batches, b"elections", Elections);
```
alphabetical for my ocd, plz

shawntabrizi commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -0,0 +1,453 @@
+type Lookup<T> = <<T as frame_system::Trait>::Lookup as StaticLookup>::Source;
+
+/// grab a new account
+fn account<T: Trait>(name: &'static str, index: u32) -> T::AccountId {
```
Please use the `account` function provided by the runtime benchmarking library, and specifically if it is being used to generate the "caller" account, use `account("caller", 0, SEED)` so we can easily track the account

shawntabrizi commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -0,0 +1,453 @@
+
+/// grab a new account
+fn account<T: Trait>(name: &'static str, index: u32) -> T::AccountId {
+	let entropy = (name, index).using_encoded(blake2_256);
```
I am asking you to remove this function, so not as relevant, but if you kept it around, I would probably tell you to use `T::Hashing` so you dont import another library here? Then you dont need sp-io or whatever

shawntabrizi commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -0,0 +1,453 @@
+		// create a bunch of candidates.
+		let all_candidates = submit_candidates::<T>(MAXIMUM_VOTE as u32, "candidates")?;
+
+		let caller = endowed_account::<T>("caller", SEED);
```
please be sure this ends up being `account("caller", 0, 0)` when using the function from the benchmarking library

shawntabrizi commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -0,0 +1,453 @@
+		// vote first `x` ones.
+		let votes = all_candidates.into_iter().take(x as usize).collect();
+
+	}: _(RawOrigin::Signed(caller), votes, stake)
```
Verify?

shawntabrizi commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -0,0 +1,453 @@
+	}: _(RawOrigin::Signed(caller))
+
+	report_defunct_voter_correct {
+		// number fo already existing members or runners
```
```suggestion
		// number of already existing members or runners
```

shawntabrizi commented:
@kianenigma only a couple of small nits. I made a quick PR for you, which if you are happy, im happy with all of this.

kianenigma commented:
## Analysis 

I ran all benchmarks on my own machine and here are the latest outputs + how I would interpret them.  

everything executed with 

```
substrate benchmark --db-cache 1024 --execution=wasm --wasm-execution=compiled --pallet elections --repeat 10 --extrinsic xxx 
```

<details>
<summary>vote: The number candidates to vote for is not significant even though we have to write it to state because the limit is 16 candidates. Likewise, the number of previous locks has a very small impact percentage. I think we should fix both to some average and use that value as the final weight and deem the weight to be Fixed. For voters we can use the maximum (16). For locks I am not sure. </summary>
<pre>
Pallet: "elections", Extrinsic: "vote", Lowest values: [], Highest values: [], Steps: [], Repeat: 10
Median Slopes Analysis
========

Model:
Time ~=    75.14
    + c        0
    + l      0.5
              µs

Min Squares Analysis
========

Data points distribution:
    c     l   mean µs  sigma µs       %
    1    20     88.66     0.942    1.0%
    2    20      85.5     0.763    0.8%
    3    20        88     1.414    1.6%
    4    20      85.5     0.763    0.8%
    5    20     84.33     1.247    1.4%
    6    20        82         0    0.0%
    7    20     81.16     0.897    1.1%
    8    20     81.16     0.897    1.1%
    9    20      82.5     0.763    0.9%
   10    20     83.33     0.471    0.5%
   11    20     82.16     0.372    0.4%
   12    20      81.5     1.607    1.9%
   13    20     81.83     1.462    1.7%
   14    20     85.33      1.49    1.7%
   15    20      83.5     1.118    1.3%
   16     1     76.33     0.745    0.9%
   16     2     77.66     0.471    0.6%
   16     3     77.66     0.471    0.6%
   16     4      76.5       0.5    0.6%
   16     5      79.5     0.763    0.9%
   16     6        80     1.527    1.9%
   16     7     83.83     0.372    0.4%
   16     8      87.5     1.118    1.2%
   16     9      83.5     0.763    0.9%
   16    10        82         0    0.0%
   16    11        81         0    0.0%
   16    12     80.33     0.471    0.5%
   16    13     80.83     0.897    1.1%
   16    14     83.16     0.897    1.0%
   16    15     81.33     0.745    0.9%
   16    16     83.33     0.471    0.5%
   16    17        85     0.816    0.9%
   16    18     88.83     1.674    1.8%
   16    19     88.83     0.372    0.4%
   16    20      86.8     2.315    2.6%

Quality and confidence:
param     error
c         0.048
l         0.036

Model:
Time ~=     77.3
    + c    0.044
    + l    0.347
              µs
</pre>
</details>

<details>
<summary>remove_voter: Similar to Vote</summary>
<pre>
Pallet: "elections", Extrinsic: "remove_voter", Lowest values: [], Highest values: [], Steps: [], Repeat: 10
Median Slopes Analysis
========

Model:
Time ~=       57
    + c        0
    + l        0
              µs

Min Squares Analysis
========

Data points distribution:
    c     l   mean µs  sigma µs       %
    1    20     63.16     1.213    1.9%
    2    20     66.33     4.678    7.0%
    3    20     59.33     1.598    2.6%
    4    20     56.66     1.105    1.9%
    5    20      56.5     1.118    1.9%
    6    20      55.5       0.5    0.9%
    7    20     55.66     0.471    0.8%
    8    20     56.16     0.372    0.6%
    9    20     58.16     1.067    1.8%
   10    20     56.83     0.897    1.5%
   11    20        62     3.366    5.4%
   12    20      57.5     0.957    1.6%
   13    20        59     2.449    4.1%
   14    20     57.83     1.343    2.3%
   15    20     56.66     0.471    0.8%
   16     1     54.33     0.471    0.8%
   16     2        56         0    0.0%
   16     3        55         1    1.8%
   16     4      55.5     0.763    1.3%
   16     5        60     3.162    5.2%
   16     6      61.5     0.763    1.2%
   16     7     55.83     1.771    3.1%
   16     8     55.33     0.745    1.3%
   16     9      55.5     0.763    1.3%
   16    10     54.83     0.372    0.6%
   16    11     55.83     1.863    3.3%
   16    12        56         1    1.7%
   16    13      56.5       1.5    2.6%
   16    14     55.33     0.471    0.8%
   16    15     56.66     1.105    1.9%
   16    16     56.66     1.972    3.4%
   16    17     55.33     0.471    0.8%
   16    18     57.33     0.942    1.6%
   16    19     56.16     0.372    0.6%
   16    20      56.4     0.663    1.1%

Quality and confidence:
param     error
c         0.046
l         0.035

Model:
Time ~=    60.55
    + c        0
    + l    0.003
              µs
</pre>
</details>

<details>
<summary>report_defunct_voter_correct: This does a binary search over `[members, candidates]` and a linear search over `runners-up`. Members and runners-up are bounded by the runtime to be very small and are insignificant. The number of already existing candidates could grow very large in principle, but note that it is not a one time cost that you can pay and then slow down the chain for ever. Candidates in this module are always short-lived. You become candidate, you either win and become a member/runner-up or you get slashed and kicked by the next round. The number of votes also gives a sub-linear growth. Final opinion: If we want to be pedantic and if the changes are worth it, the call should submit the number of candidates, and number of votes of the reported account, so can compute weights from it. Else, fixed weight should also be okay, based on 1- votes are end of the day limited to 16. Candidates are end of the day limited to one term and then we clean them all. </summary>
<pre>
Pallet: "elections", Extrinsic: "report_defunct_voter_correct", Lowest values: [], Highest values: [], Steps: [], Repeat: 10
Median Slopes Analysis
========

Model:
Time ~=        0
    + m     3.09
    + c    1.822
    + v    22.28
              µs

Min Squares Analysis
========

Data points distribution:
    m     c     v   mean µs  sigma µs       %
    1   100    16     430.6      5.99    1.3%
    2   100    16       451     9.521    2.1%
    3   100    16     437.1     5.927    1.3%
    4   100    16       470     4.725    1.0%
    5   100    16     438.3     1.247    0.2%
    6   100    16     440.5      2.14    0.4%
    7   100    16     443.1     2.266    0.5%
    8   100    16     444.8     2.114    0.4%
    9   100    16     448.3     2.426    0.5%
   10   100    16     468.5     11.57    2.4%
   11   100    16     453.1     3.715    0.8%
   12   100    16     479.5     10.09    2.1%
   13   100    16       457     1.825    0.3%
   14   100    16       476     5.859    1.2%
   15   100    16     482.3     5.647    1.1%
   16   100    16     505.5     8.057    1.5%
   17   100    16       514     7.745    1.5%
   18   100    16     492.5     11.44    2.3%
   19   100    16     493.1     13.19    2.6%
   20     1    16     304.3     3.091    1.0%
   20    10    16     321.5     1.258    0.3%
   20    19    16     339.8     1.674    0.4%
   20    28    16       354     1.414    0.3%
   20    37    16     370.6     3.496    0.9%
   20    46    16       386     1.154    0.2%
   20    55    16     401.1     0.687    0.1%
   20    64    16     419.3     0.942    0.2%
   20    73    16     459.6     12.69    2.7%
   20    82    16       483      9.66    2.0%
   20    91    16       473     11.35    2.4%
   20   100     1       147     3.265    2.2%
   20   100     2     178.3     4.921    2.7%
   20   100     3     202.3     4.027    1.9%
   20   100     4     216.1     1.213    0.5%
   20   100     5     235.1     1.462    0.6%
   20   100     6       270     10.01    3.7%
   20   100     7     273.6     0.745    0.2%
   20   100     8     300.6     5.962    1.9%
   20   100     9       329     2.768    0.8%
   20   100    10     371.1     7.197    1.9%
   20   100    11       370     3.696    0.9%
   20   100    12     383.6     2.624    0.6%
   20   100    13     440.6      9.86    2.2%
   20   100    14     437.3     6.749    1.5%
   20   100    15     465.5       5.5    1.1%
   20   100    16     485.5     4.911    1.0%

Quality and confidence:
param     error
m         0.156
c         0.033
v         0.208

Model:
Time ~=        0
    + m    3.189
    + c    1.931
    + v    23.16
              µs</pre>
</details>

<details>
<summary>report_defunct_voter_incorrect: incorrect report is typically cheaper than a correct report, since it means that one of the searches above terminated earlier than expected. The benchmark is designed worse case: the reported account votes for `v-1` invalid accounts and then finally a valid one, which makes the caller get slashed and the call to fail. The worse case of this call is theoretically the same as the previous one, where we have to search all 3 vectors (members, runners-up and candidates).</summary>
<pre>
Pallet: "elections", Extrinsic: "report_defunct_voter_incorrect", Lowest values: [], Highest values: [], Steps: [], Repeat: 10
Median Slopes Analysis
========

Model:
Time ~=        0
    + m    2.875
    + c    1.638
    + v     23.5
              µs

Min Squares Analysis
========

Data points distribution:
    m     c     v   mean µs  sigma µs       %
    1   100    16     466.6     4.307    0.9%
    2   100    16     459.5      11.2    2.4%
    3   100    16       475     3.265    0.6%
    4   100    16     483.3     3.248    0.6%
    5   100    16     474.1     4.374    0.9%
    6   100    16     492.3      3.59    0.7%
    7   100    16     489.8     2.034    0.4%
    8   100    16       491     6.164    1.2%
    9   100    16     488.1     4.879    0.9%
   10   100    16     495.3     2.426    0.4%
   11   100    16       500     2.943    0.5%
   12   100    16       494     10.56    2.1%
   13   100    16     507.3     4.642    0.9%
   14   100    16     508.1     10.07    1.9%
   15   100    16     512.3     8.919    1.7%
   16   100    16     518.3     2.211    0.4%
   17   100    16     515.5     9.995    1.9%
   18   100    16       514     9.609    1.8%
   19   100    16     540.5     3.403    0.6%
   20     1    16     333.8     5.727    1.7%
   20    10    16     348.8     5.112    1.4%
   20    19    16     353.8     6.517    1.8%
   20    28    16     355.1     5.927    1.6%
   20    37    16     375.3     6.289    1.6%
   20    46    16     387.5     0.763    0.1%
   20    55    16       403     0.816    0.2%
   20    64    16     421.1     1.067    0.2%
   20    73    16     438.3     3.399    0.7%
   20    82    16     447.1     1.343    0.3%
   20    91    16     465.8     4.374    0.9%
   20   100     2     183.3     5.374    2.9%
   20   100     3       205     7.527    3.6%
   20   100     4     239.1     2.409    1.0%
   20   100     5     259.8      3.76    1.4%
   20   100     6     266.3     2.924    1.0%
   20   100     7     307.6     4.784    1.5%
   20   100     8     331.8      2.91    0.8%
   20   100     9     360.8     3.435    0.9%
   20   100    10     386.3     1.972    0.5%
   20   100    11     409.1     4.139    1.0%
   20   100    12     419.5     11.58    2.7%
   20   100    13     461.3     0.942    0.2%
   20   100    14     460.3      13.8    2.9%
   20   100    15     501.3     11.67    2.3%
   20   100    16     494.8     19.77    3.9%

Quality and confidence:
param     error
m         0.184
c         0.039
v         0.267

Model:
Time ~=        0
    + m    1.872
    + c    2.013
    + v    22.82
              µs
</pre>
</details>

<details>
<summary>submit_candidacy: Similar to the above ones, we need to check that a candidate is not a duplicate with either of `members`, `runners-up` and `candidates`. Similarly, the former two are not significant. The third is subject to the same argument: either fix it, or ask the user to provide the length of the candidates that need to be searched logarithmically</summary>
<pre>
Pallet: "elections", Extrinsic: "submit_candidacy", Lowest values: [], Highest values: [], Steps: [], Repeat: 10
Median Slopes Analysis
========

Model:
Time ~=    38.51
    + m     0.25
    + c    0.269
              µs

Min Squares Analysis
========

Data points distribution:
    m     c   mean µs  sigma µs       %
    1   100     66.83     0.687    1.0%
    2   100     66.33     0.471    0.7%
    3   100     66.16     0.372    0.5%
    4   100     66.66     0.471    0.7%
    5   100        68         0    0.0%
    6   100      66.5       0.5    0.7%
    7   100        67         0    0.0%
    8   100     66.83     0.372    0.5%
    9   100        67         0    0.0%
   10   100        67         0    0.0%
   11   100        69     0.577    0.8%
   12   100     69.16     0.372    0.5%
   13   100      68.5     1.118    1.6%
   14   100     68.66     0.471    0.6%
   15   100        69         0    0.0%
   16   100     68.83     0.372    0.5%
   17   100     69.66     0.471    0.6%
   18   100      78.5     6.551    8.3%
   19   100     74.16     0.687    0.9%
   20     1        43         0    0.0%
   20    10     46.66     0.471    1.0%
   20    19     50.16     0.372    0.7%
   20    28      52.5     0.763    1.4%
   20    37     53.16     0.372    0.6%
   20    46        56         0    0.0%
   20    55     58.33     0.471    0.8%
   20    64     60.83     0.687    1.1%
   20    73     65.83     0.687    1.0%
   20    82     66.66     0.471    0.7%
   20    91        68     0.816    1.2%
   20   100      70.3     0.458    0.6%

Quality and confidence:
param     error
m         0.028
c         0.006

Model:
Time ~=    36.85
    + m    0.336
    + c    0.282
              µs
</pre>
</details>

<details>
<summary>renounce_candidacy_candidate: a candidate that wants to renounce will first check to be a member or runner, which is not the case here, and is insignificant because number `m` is bounded by 20 in our runtime. Then we do log search in candidates list. The same argument applies. We could add an enum to this call so that the renouncing origin will say: they are either of: enum Renouncing { Member, RunneUp, Candidate(usize) }  so we know exactly who wants to renounce. </summary>
<pre>
allet: "elections", Extrinsic: "renounce_candidacy_candidate", Lowest values: [], Highest values: [], Steps: [], Repeat: 10
Median Slopes Analysis
========

Model:
Time ~=    30.39
    + m    0.363
    + c    0.158
              µs

Min Squares Analysis
========

Data points distribution:
    m     c   mean µs  sigma µs       %
    1   100      48.5     0.763    1.5%
    2   100     46.16     0.372    0.8%
    3   100     46.83     1.067    2.2%
    4   100        49     0.577    1.1%
    5   100     48.16     0.372    0.7%
    6   100     48.66     0.745    1.5%
    7   100     53.16     3.077    5.7%
    8   100     48.16     0.372    0.7%
    9   100      49.5       0.5    1.0%
   10   100     49.66     0.471    0.9%
   11   100      49.5       0.5    1.0%
   12   100        50         0    0.0%
   13   100     50.16     0.372    0.7%
   14   100     50.66     0.471    0.9%
   15   100     53.16     0.897    1.6%
   16   100     52.16     0.372    0.7%
   17   100     60.33     5.962    9.8%
   18   100     63.66     0.471    0.7%
   19   100     56.33     2.624    4.6%
   20     1        37         0    0.0%
   20    10     38.83     0.372    0.9%
   20    19     41.16     0.372    0.9%
   20    28      42.5       0.5    1.1%
   20    37      44.5     2.986    6.7%
   20    46     43.16     0.372    0.8%
   20    55        51     12.09   23.7%
   20    64      47.5       0.5    1.0%
   20    73        50         0    0.0%
   20    82      50.5       0.5    0.9%
   20    91     51.16     0.372    0.7%
   20   100      53.5      0.67    1.2%

Quality and confidence:
param     error
m         0.049
c          0.01

Model:
Time ~=    27.34
    + m    0.482
    + c    0.187
              µs
</pre>
</details>

<details>
<summary>renounce_candidacy_member_runner_up: as predicted above, the number candidates has almost no impact. a very small change with the number of members and runners-up.</summary>
<pre>
Pallet: "elections", Extrinsic: "renounce_candidacy_member_runner_up", Lowest values: [], Highest values: [], Steps: [], Repeat: 10
Median Slopes Analysis
========

Model:
Time ~=    64.75
    + c        0
    + m    0.333
              µs

Min Squares Analysis
========

Data points distribution:
    c     m   mean µs  sigma µs       %
    1    20        72     3.162    4.3%
   10    20     69.66     0.471    0.6%
   19    20        72     0.816    1.1%
   28    20     72.16     0.687    0.9%
   37    20     72.16     1.462    2.0%
   46    20     70.33     0.471    0.6%
   55    20        68         0    0.0%
   64    20        67         0    0.0%
   73    20     67.16     0.372    0.5%
   82    20      69.5       1.5    2.1%
   91    20     75.16     4.412    5.8%
  100     1     63.83     0.687    1.0%
  100     2        66         0    0.0%
  100     3     62.33     0.471    0.7%
  100     4     62.83     0.372    0.5%
  100     5        63         0    0.0%
  100     6        64         0    0.0%
  100     7      64.5     0.763    1.1%
  100     8      64.5       0.5    0.7%
  100     9     64.83     0.372    0.5%
  100    10     65.16     0.897    1.3%
  100    11      65.5     0.763    1.1%
  100    12     65.83     0.372    0.5%
  100    13     67.66     0.745    1.1%
  100    14     67.83     0.372    0.5%
  100    15     70.33     2.687    3.8%
  100    16        69     0.577    0.8%
  100    17        68     0.816    1.2%
  100    18      67.5     0.763    1.1%
  100    19     67.33     0.471    0.6%
  100    20      70.5     1.746    2.4%

Quality and confidence:
param     error
c         0.006
m         0.029

Model:
Time ~=    63.96
    + c        0
    + m    0.362
              µs
</pre>
</details>

<details>
<summary>remove_member_without_replacement: this triggers phragmen internally. For now I made it variable over only the number of candidates `c` with at most 500 voters voting for 16 of the candidates (same setup as on_initialize). The times are highe but not really a function of the number of candidates. Phragmen is O(edge_count), which is mostly dominant by voters. I will do a separate benchmark for this. </summary>
<pre>
Pallet: "elections", Extrinsic: "remove_member_without_replacement", Lowest values: [], Highest values: [], Steps: [], Repeat: 10
Median Slopes Analysis
========

Model:
Time ~=    57180
    + c      393
              µs

Min Squares Analysis
========

Data points distribution:
    c   mean µs  sigma µs       %
    1     16030     87.08    0.5%
   10     35860     498.8    1.3%
   19     63180      1208    1.9%
   28     76840      1546    2.0%
   37     80700      2217    2.7%
   46     81440     991.5    1.2%
   55     84490      1761    2.0%
   64     89400      2501    2.7%
   73     87370      3549    4.0%
   82     83320      1160    1.3%
   91     85840      3015    3.5%
  100     83340      1417    1.7%

Quality and confidence:
param     error
c         54.41

Model:
Time ~=    44590
    + c    548.9
              µs
</pre>
</details>


<details>
<summary>remove_member_with_replacement: this one is dead cheap. Number of already existing members/runners has no tangible impact. Needs a fixed weight.</summary>
<pre>
Pallet: "elections", Extrinsic: "remove_member_with_replacement", Lowest values: [], Highest values: [], Steps: [], Repeat: 10
Median Slopes Analysis
========

Model:
Time ~=     72.4
    + m      0.2
              µs

Min Squares Analysis
========

Data points distribution:
    m   mean µs  sigma µs       %
   14        79     4.546    5.7%
   15     70.33     0.745    1.0%
   16     72.83     0.687    0.9%
   17      76.5       0.5    0.6%
   18     76.33     0.471    0.6%
   19     77.16     0.372    0.4%
   20     73.16     2.733    3.7%

Quality and confidence:
param     error
m         0.274

Model:
Time ~=    75.25
    + m        0
              µs
</pre>
</details>


<details>
<summary>on_initialize: As expected, almost same numbers as remove_member_without_replacement</summary>
<pre>
Pallet: "elections", Extrinsic: "on_initialize", Lowest values: [], Highest values: [], Steps: [], Repeat: 10
Median Slopes Analysis
========

Model:
Time ~=    49550
    + c    391.8
              µs

Min Squares Analysis
========

Data points distribution:
    c   mean µs  sigma µs       %
    1     12810     284.1    2.2%
   10     31980      1623    5.0%
   19     54580      1264    2.3%
   28     68020      2498    3.6%
   37     70110      1068    1.5%
   46     71240     366.5    0.5%
   55     73660      1299    1.7%
   64     75260     879.7    1.1%
   73     75680     671.8    0.8%
   82     77080      1158    1.5%
   91     84730      2575    3.0%
  100     81180      1615    1.9%

Quality and confidence:
param     error
c         42.78

Model:
Time ~=    36430
    + c    559.6
              µs
</pre>
</details>

<details>
<summary>phragmen: This is a new benchmark that I added just to observe how phragmen works within assumed boundaries of this module: DesiredMembers being 13, 7 runners-up and a range of candidates and voters of at 100 and 500 respectively. As already know, phragmen is linear with respect to number of edges. The proposal here is to make the weight maximum. The only situations where we call phragmen in a block is then at the end of the term (every 24 hrs or week) and when we remove a member with no replacement, which is very rare + it is a root call </summary>
<pre>
246    16     40950     190.8    0.4%
  100   295    16     47600     867.9    1.8%
  100   344    16     54460     477.5    0.8%
  100   393    16     60970     261.9    0.4%
  100   442    16     69990     925.4    1.3%
  100   491    16     75060     480.3    0.6%
</pre>
</details>

gavofyork commented:
Given that candidates are unbounded we probably and that I/O at least is `O(N)` (even if search is `O(log N)`), we sholud proably play it safe and have the user pass in an upper bound for `N` as a parameter as needed.

gavofyork commented:
We could also bound candidates to something sensible (e.g. 1000), but then we might also need to have a sliding scale for the deposit (e.g. 2x every 100) so that it's not easily griefable.

kianenigma commented:
> we sholud proably play it safe and have the user pass in an upper bound for N as a parameter as needed.

For now this is the approach taken here.

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -258,46 +300,72 @@ decl_module! {
-		/// Writes: O(V) given `V` votes. V is bounded by 16.
+		/// Base weight: 50us
+		/// State reads:
+		/// 	- Candidates.len() + Members.len() + RunnersUp.len()
+		/// 	- Voting (is_voter)
```
So as discussed, I assume that reading just the length of some storage is different than reading it later. 

Correct me if I am wrong.

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -334,26 +407,56 @@ decl_module! {
+		///     longer a candidate nor an active member or a runner-up.
+		///
+		///
+		/// The origin must provide the number of current candidates for the purpose of accurate
+		/// weight calculation.
```
```suggestion
		/// The origin must provide the number of current candidates and votes of the reported target 
		/// for the purpose of accurate weight calculation.
```

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -334,26 +407,56 @@ decl_module! {
+		/// 	- Lock(reporter || target)
+		/// 	- AccountBalance(reporter) + AccountBalance(target)
+		/// 	- Voting(reporter || target)
+		/// Note: the db access is worse with respect to db, which is when the report is correct.
```
There is seriously no way to know if the report is correct or not without executing the whole damn thing, hence we assume worse case here. The diff is merely one db read extra.

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -334,26 +407,56 @@ decl_module! {
+		)]
+		fn report_defunct_voter(
+			origin,
+			defunct: DefunctVoter<<T::Lookup as StaticLookup>::Source>,
```
@shawntabrizi I recall that you once asked about whether if this is needed anymore. this can be swapped with just T::AccountId as well. Not sure if source and lookup is needed here.

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -334,26 +407,56 @@ decl_module! {
-		/// #### State
-		/// Reads: O(NLogM) given M current candidates and N votes for `target`.
-		/// Writes: O(1)
+		/// No Base weight based on min square analysis.
```
```suggestion
		/// No Base weight based on min square analysis.
		/// Complexity of candidate_count: 1.755 us
		/// Complexity of vote_count: 17.97us
```

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -406,55 +532,97 @@ decl_module! {
 		/// - `origin` is a current member. In this case, the bond is unreserved and origin is
 		///   removed as a member, consequently not being a candidate for the next round anymore.
 		///   Similar to [`remove_voter`], if replacement runners exists, they are immediately used.
-		#[weight = (2_000_000_000, DispatchClass::Operational)]
```
So I don't know why this was operational before? I think it was a mistake.

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -406,55 +532,97 @@ decl_module! {
+		/// 		- RunnersUp (remove_and_replace_member),
+		/// 		- AccountData(who) (unreserve)
+		///
+		/// TODO: and calls into ChangeMembers??
```
@apopiak perhaps you can help here. Should this be noted? or I should wait for you PR for now?

apopiak replied:
These definitely need to be factored in.
For Polkadot we know that it will be using `collective` under the hood, so you can cross-validate benchmark results for this function with [the annotations in the `collective` implementation here](https://github.com/paritytech/substrate/pull/5802/files#diff-fd8e3211907a8723b36fbf8504230548R771-R787).

apopiak replied:
Your benchmarks should tell you all you need to know. The collective docs are just for cross-checking.

kianenigma replied:
As discussed offline, this will be addressed later.

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -466,34 +634,73 @@ decl_module! {
+					// lot of weight. just abort.
+					return Err(Error::<T>::InvalidReplacement.into());
+				}
+				(true, false) => {
```
@athei just so you can see it, I think it is all fine though.

athei replied:
Yes looks good.

kianenigma commented:
Another TOOD: 

- [ ] I am counting in db reads and writes into account data of origin, while I should NOT.

athei commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -466,34 +634,73 @@ decl_module! {
+				(false, true) => {
+					// prediction was that we will have a replacement, so we don't charge a whole
+					// lot of weight. just abort.
+					return Err(Error::<T>::InvalidReplacement.into());
```
Why not refunding here as in the `(true, false)` case? You are charging for db accesses that are never executed.

kianenigma replied:
Right, that would be an extra benefit. The other one that I refund now is way more severe because it charges the full block. Fixed.

athei commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -466,34 +634,73 @@ decl_module! {
+						// refund to the variant where we have a replacement. This still an
+						// overestimate but fine for now.
+						48 * WEIGHT_PER_MICROS + T::DbWeight::get().reads_writes(1, 0)
+					).into());
```
You can lose that `.into()`. `with_weight` already returns the correct type AFAIK.

thiolliere commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -334,26 +407,58 @@ decl_module! {
+			// Optimisation Note: We read this again in is_defunct_voter, still duplicate overlay
+			// read.
+			ensure!(
+				<Voting<T>>::get(&target).1.len() as u32 <= vote_count,
```
checking this length involve decoding it, and decoding is considered `O(lenght)` but not sure of the real cost actually, in democracy I did swap the tuple element and implemented a manual decoding. https://github.com/paritytech/substrate/pull/5828#discussion_r419484688

If wanted we should release a new codec version with https://github.com/paritytech/parity-scale-codec/pull/209

Otherwise depending on the constaint here maybe it fine.

kianenigma replied:
Swapping will not work here, what will work is decoupling voting targets and balances. But it used to like this and we merged them because we always read these two together. 

For now I will leave it and have a note there to maybe improve later.

kianenigma replied:
actually I can re-org the functions in a way that I read the voters once and then pass them around. this should be a bit more expensive in-memory, but will prevent the need for multiple decoding. 

@shawntabrizi technically this might need a re-run :( but I think we will do this at the very end once more. So not a big deal for now.

thiolliere replied:
ok, also I see now that votes is of lenght 16 as maximum so fine.

kianenigma commented:
If we will re-visit the usage of saturating later, then this is done as far as I can see. 

@apopiak your comment about ChangeMembers will be addressed later. 
@thiolliere your comment is addressed and I re-ordered the code a bit to make it only read once and pass the data around to the function.

shawntabrizi commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -258,46 +297,67 @@ decl_module! {
+			origin,
+			votes: Vec<T::AccountId>,
+			#[compact] value: BalanceOf<T>,
+			would_create: bool,
```
big no here imo!

You have introduced some parameter to a function to save on 10 microseconds?

This is def not right

kianenigma replied:
yeah I was perhaps trying to be too pedantic here. You're right this is not significant and will revert it.

shawntabrizi commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -334,27 +399,62 @@ decl_module! {
-		#[weight = 1_000_000_000]
-		fn report_defunct_voter(origin, target: <T::Lookup as StaticLookup>::Source) {
+		#[weight =
+			defunct.candidate_count as Weight * (2 * WEIGHT_PER_MICROS) +
```
This MUST be saturating before we can merge.

shawntabrizi commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -380,21 +479,40 @@ decl_module! {
-		fn submit_candidacy(origin) {
+		#[weight =
+			35 * WEIGHT_PER_MICROS +
+			*candidate_count as Weight * (375 * WEIGHT_PER_NANOS) +
```
this MUST be saturating before we can merge

thiolliere commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -334,27 +399,62 @@ decl_module! {
+		///  	- Voting(reporter)
+		///  	- Candidate.len()
+		///  	- Voting(Target)
+		///  	- Candidates, Members, RunnersUp (is_defunct_voter)
```
this does not only reads 3 values but at most 3*defunct.vote_count

kianenigma replied:
can you explain a bit more: under the hood, we are searching inside these 3 storage items, all of which are `StorageValue` and contain a vector, so from the standpoint of just _count_ of storage reads, I think it is 3.

thiolliere replied:
ah you are right, I missread.

shawntabrizi commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -406,55 +524,93 @@ decl_module! {
-
-			Err(Error::<T>::InvalidOrigin)?
+		}]
+		fn renounce_candidacy(origin, renouncing: Renouncing) {
```
This would be 3 functions if I were designing the API

shawntabrizi replied:
It would not require this renouncing type for example, and would reduce the size of the tx by removing one parameter

kianenigma replied:
Store for another day?

shawntabrizi replied:
indeed

shawntabrizi commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -406,55 +524,93 @@ decl_module! {
+		#[weight =  match *renouncing {
+			Renouncing::Candidate(count) => {
+				18 * WEIGHT_PER_MICROS +
+				(count as Weight) * 235 * WEIGHT_PER_NANOS +
```
Must be saturating before we can merge

kianenigma commented:
Fair points, addressed.

thiolliere commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -466,34 +622,57 @@ decl_module! {
 
 				if !had_replacement {
+					// if we end up here, we will charge a full block weight.
 					Self::do_phragmen();
```
where is the full block charged ?
